rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -44,38 +44,36 @@ void Recorder::record(const RecordOptions & record_options)
 
   std::future<void> discovery_future;
   if (!record_options.is_discovery_disabled) {
-    discovery_future = launch_topics_discovery(
+    auto launch_discovery = std::bind(
+      &Recorder::launch_topics_discovery, this,
       record_options.topic_polling_interval, record_options.topics);
+    discovery_future = std::async(std::launch::async, launch_discovery);
   }
 
   record_messages();
 
-  if (!record_options.is_discovery_disabled) {
+  if (discovery_future.valid()) {
     discovery_future.wait();
   }
+
   subscriptions_.clear();
 }
 
-std::future<void> Recorder::launch_topics_discovery(
+void Recorder::launch_topics_discovery(
```

Comment:
If you move the `async` call out the function should no longer have launch in its name.